File: 1030-smallest-string-starting-from-leaf/smallest-string-starting-from-leaf.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// Definition for a binary tree node.
class Solution {
    public String smallestFromLeaf(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        traverse(root, path, res);
        
        List<String> strings = new ArrayList<>();
        for (List<Integer> list : res) {
            Collections.reverse(list);
            StringBuilder sb = new StringBuilder();
            for (int val : list) {
                sb.append((char)(val + 'a'));
            }
            strings.add(sb.toString());
        }
        
        Collections.sort(strings);
        return strings.get(0);
    }
    
    private void traverse(TreeNode node, List<Integer> path, List<List<Integer>> res) {
        if (node == null) {
            return;
        }
        
        path.add(node.val);
        if (node.left == null && node.right == null) {
            res.add(new ArrayList<>(path));
        } else {
            traverse(node.left, path, res);
            traverse(node.right, path, res);
        }
        path.remove(path.size() - 1);
    }
}

File: 988-smallest-string-starting-from-leaf/smallest-string-starting-from-leaf.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root):
        s = []
        res = []
        self.traverse(root, [], res)
        
        for i in res:
            i.reverse()
            x = []
            for j in i:
                y = chr(j + ord('a'))
                x.append(y)
            string = ''.join(x)
            
            if len(s) == 0 or string < ''.join(s):
                s = list(string)
        
        return ''.join(s)
    
    def traverse(self, root, lst, res):
        if not root:
            return
        lst.append(root.val)
        
        if not root.left and not root.right:
            res.append(list(lst))
            return
        self.traverse(root.left, list(lst), res)
        self.traverse(root.right, list(lst), res)
        
        lst.pop()     
